src/plugins/core/filters.ts

```diff
@@ -0,0 +1,333 @@
+    switch (cmd.type) {
+      case "CREATE_SHEET":
+        this.history.update("tables", cmd.sheetId, {});
+        break;
+      case "DELETE_SHEET":
+        const filterTables = { ...this.tables };
+        delete filterTables[cmd.sheetId];
+        this.history.update("tables", filterTables);
+        break;
+      case "DUPLICATE_SHEET":
+        this.history.update("tables", cmd.sheetIdTo, deepCopy(this.tables[cmd.sheetId]));
+        break;
+      case "ADD_COLUMNS_ROWS":
+        this.onAddColumnsRows(cmd);
+        break;
+      case "REMOVE_COLUMNS_ROWS":
+        this.onDeleteColumnsRows(cmd);
+        break;
+      case "CREATE_FILTER_TABLE": {
+        const zone = union(...cmd.target);
+        const newFilterTable = this.createFilterTable(zone);
+        this.history.update("tables", cmd.sheetId, newFilterTable.id, newFilterTable);
+        break;
+      }
+      case "REMOVE_FILTER_TABLE": {
+        const tables: Record<UID, FilterTable> = {};
+        for (const filterTable of this.getFilterTables(cmd.sheetId)) {
+          if (cmd.target.every((zone) => !intersection(zone, filterTable.zone))) {
+            tables[filterTable.id] = filterTable;
+          }
+        }
+        this.history.update("tables", cmd.sheetId, tables);
+        break;
+      }
+      case "UPDATE_CELL": {
+        this.onUpdateCell(cmd);
+      }
+    }
+  }
+
+  getFilters(sheetId: UID): Filter[] {
+    return this.getFilterTables(sheetId)
+      .map((filterTable) => filterTable.filters)
+      .flat();
+  }
+
+  getFilterTables(sheetId: UID): FilterTable[] {
+    return this.tables[sheetId] ? Object.values(this.tables[sheetId]).filter(isDefined) : [];
+  }
+
+  getFilter(sheetId: UID, col: number, row: number): Filter | undefined {
+    return this.getFilterTable(sheetId, col, row)?.filters.find((filter) => filter.col === col);
+  }
+
+  getFilterId(sheetId: UID, col: number, row: number): FilterId | undefined {
+    return this.getFilter(sheetId, col, row)?.id;
+  }
+
+  getFilterTable(sheetId: UID, col: number, row: number): FilterTable | undefined {
+    return this.getFilterTables(sheetId).find((filterTable) =>
+      isInside(col, row, filterTable.zone)
+    );
+  }
+
+  /** Get the filter tables that are fully inside the given zone */
+  getFilterTablesInZone(sheetId: UID, zone: Zone): FilterTable[] {
+    return this.getFilterTables(sheetId).filter((filterTable) =>
+      isZoneInside(filterTable.zone, zone)
+    );
+  }
+
+  doesZonesContainFilter(sheetId: UID, zones: Zone[]): boolean {
+    for (const zone of zones) {
+      for (const filterTable of this.getFilterTables(sheetId)) {
+        if (intersection(zone, filterTable.zone)) {
+          return true;
+        }
+      }
+    }
+    return false;
+  }
+
+  private onAddColumnsRows(cmd: AddColumnsRowsCommand) {
+    for (const filterTable of this.getFilterTables(cmd.sheetId)) {
+      if (!filterTable) {
+        return;
+      }
+      const zone = expandZoneOnInsertion(
```

Comment:
I originally wanted to use ranges, but it doesn't work that well.

One issue is that every time a row is added/removed in the filter table, I have to add/remove items in the `filters[]` array of the table. So after adapting a range, I have to check if new cols were added/removed (I don't even know what command triggered the change in the adapt ranges), then rebuild the filters array.

The other issue is that I'm pretty sure I just cannot use ranges inside `Filters` atm, because inserting a cell above a table column would shift is down, which we really don't want. Or we'd need additional logic "only update Filter range if the FilterTabel range was updated".

In the end, I think using ranges could be done, but wouldn't really improve the code.